File: src/main/scala/kim/dencka/scala/easy/GuessNumberHigherOrLower.scala
package kim.dencka.scala.easy

// KIM_INFO IT SOLVED

class GuessNumberHigherOrLower {
  import GuessNumberHigherOrLower._

  def guessNumber(n: Int): Int = {
    @annotation.tailrec
    def run(low: Int, high: Int): Int = {
      val middle = low + (high - low) / 2

      guess(middle) match {
        case -1 => run(low, middle - 1)
        case 1  => run(middle + 1, high)
        case 0  => middle
      }
    }
    run(1, n)
  }
}

object GuessNumberHigherOrLower {
  // -1: Your guess is higher than the number I picked(i.e.num > pick)
  // 1: Your guess is lower than the number I picked(i.e.num < pick)
  // 0: your guess is equal to the number I picked(i.e.num == pick)

  def guess(k: Int): Int = ???

}

File: src/main/scala/kim/dencka/java/easy/GuessNumberHigherOrLower.java
package kim.dencka.java.easy;

public class GuessNumberHigherOrLower {
    public static void main(String[] args) {
        System.out.println(guessNumber(2));
    }

    static int pickedNumber = 2;

    public static int guessNumber(int n) {
        int l = 1, h = n;
        while (l <= h) {
            int mid = (h + l) / 2;
            int res = guess(mid);
            if (res == 0) return mid;
            if (res == -1) {
                h = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return -1;
    }


    public static int guess(int num) {
        return Integer.compare(pickedNumber, num);
    }
}
